woocommerce/google-listings-and-ads
# Adjustments to scss build to align with wc-admin approach

## Background

One of the things we noticed early on with our wc-admin generated extension is that we needed to import (I.e duplicate) sass variables into individual components scss files ([example](https://github.com/woocommerce/google-listings-and-ads/blob/trunk/js/src/get-started-page/get-started-card/index.scss#L1-L2)) -  when working with wc-admin we;ve never needed to do that.

it was annoying me so I thought I'd figure it today.

## What does wc-admin do?

Essentially it prepends a range of variables to all the files during the build process https://github.com/woocommerce/woocommerce-admin/blob/main/webpack.config.js#L149-L153 - variables don't output anything so there is no risk of duplicating anything and it means they don't need to add imports to all the scss. You can just rely on they being available.

They then have some "global" styles that are essentially imported once into the main entry point e.g. https://github.com/woocommerce/woocommerce-admin/tree/main/client/stylesheets/shared

## What does this PR do

d534e4b removes some of the duplicated imports we currently have in place

554e62b sets up a similar structure to wc-admin for global and abstract/partial scss files (100% open to renaming/moving etc.)

6c10692 adjust the webpack build process to prepend the variables and make sure they are available to us in our components without needing to explicitly import.

One thing to note here (and I'd love if @tomalec or @ecgan have a nicer way to handle this) is that because we are using wp-scripts to handle almost all our configuration (Ideally I'd like to continue to do so for ease of maintenance - hopefully) I've tried to s[imply update the default sass-loader configuration](https://developer.wordpress.org/block-editor/packages/packages-scripts/#provide-your-own-webpack-config) that wp-scripts has (but it wasn't as easy as I'd hoped).

@ecgan @tomalec if you know a nicer and [more robust way we can simply add the prependData option](https://github.com/woocommerce/google-listings-and-ads/compare/css-import-clean?expand=1#diff-1fb26bc12ac780c7ad7325730ed09fc4c2c3d757c276c3dacc44bfe20faf166fR34-R48) I'd be open to it. (it is 100% hacky and more a proof of concept at the moment - I just didn't want to rewrite all the loader config from scratch and couldn't think of quick way to do it)

**Update** - investigated using the webpage-merge package to merge rules but it had/has some bugs. Decided to simply try a different approach ead9b07

# How to test

* Pull down this branch
* Npm install and build
* Open the Marketing > Google Tab and everything should look as expected.
* Check the built index.css file to make sure we don't have any unnecessary duplication - from what I can see the file size is the same before/after - we just don't need to keep importing variables all the time anymore.
